include/nanobind/nb_traits.h

```diff
@@ -114,6 +114,17 @@ template <template <typename> typename Op, typename Arg>
 struct detector<std::void_t<Op<Arg>>, Op, Arg>
     : std::true_type { };
 
+template <typename T, typename...>
```

Comment:
Having a dependency of a core part of nanobind on `std::variant` is unfortunately an absolute no-go for me, so something bout these traits would have to change. 

Perhaps we can start with some discussion on the PR on what specific issue they address, and how else it could be achieved?

Reply:
I understood. It was rough to include two headers globally just for signature string generation. Actually, I also had trouble with this part, and the reason why `type traits` depends on STL was to avoid changing the design of the type caster.

I pushed a new commit to discuss implementation policy. Static member variables `ArgName` are added to optional and variant casters, and meta-classes are added to `nb_func.h`. The dependency on STL header has been removed, but it might look a bit messy. What do you think?

Reply:
I noticed that I should have put the template specialization of `remove_opt_mono` I added into each header. How about this change? (0142b12)